src/main/java/io/netty/incubator/codec/http3/QpackDecoder.java

```diff
@@ -16,94 +16,383 @@
+     * unblocked and the header block is completely decoded.
+     *
+     * @param qpackDecoderStream {@link QuicStreamChannel} for the QPACK decoder stream.
+     * @param streamId for the stream on which this header block was received.
+     * @param in {@link ByteBuf} containing the header block.
+     * @param length Number of bytes to be read from {@code in}
+     * @param sink {@link BiConsumer} to
+     * @param whenDecoded {@link Runnable} to invoke when a blocked decode finishes decoding.
+     * @return {@code true} if the headers were decoded.
      */
-    public void decode(ByteBuf in, BiConsumer<CharSequence, CharSequence> sink) throws QpackException {
-        // Required Insert Count
-        // https://tools.ietf.org/html/draft-ietf-quic-qpack-19#section-4.5.1.1
-        QpackUtil.decodePrefixedInteger(in, 8);
+    public boolean decode(QuicStreamChannel qpackDecoderStream, long streamId, ByteBuf in,
+                          int length, BiConsumer<CharSequence, CharSequence> sink, Runnable whenDecoded)
+            throws QpackException {
+        final int initialReaderIdx = in.readerIndex();
+        final int requiredInsertCount = decodeRequiredInsertCount(in);
+        if (shouldWaitForDynamicTableUpdates(requiredInsertCount)) {
+            blockedStreamsCount++;
+            blockedStreams.computeIfAbsent(requiredInsertCount, __ -> new ArrayList<>(2))
+                    .add(requiredInsertCount, whenDecoded);
+            return false;
+        }
 
-        // Delta Base
-        // https://tools.ietf.org/html/draft-ietf-quic-qpack-19#section-4.5.1.2
-        QpackUtil.decodePrefixedInteger(in, 7);
+        in = in.readSlice(length - (in.readerIndex() - initialReaderIdx));
+        final int base = decodeBase(in, requiredInsertCount);
 
         while (in.isReadable()) {
             byte b = in.getByte(in.readerIndex());
-            if ((b & 0x80) == 0x80) {
-                // 1xxxxxxx
-                decodeIndexed(in, sink);
-            } else if ((b & 0xc0) == 0x40) {
-                // 01xxxxxx
-                decodeLiteralWithNameReference(in, sink);
-            } else if ((b & 0xe0) == 0x20) {
-                // 001xxxxx
+            if (isIndexed(b)) {
+                decodeIndexed(in, sink, base);
+            } else if (isIndexedWithPostBase(b)) {
+                decodeIndexedWithPostBase(in, sink, base);
+            } else if (isLiteralWithNameRef(b)) {
+                decodeLiteralWithNameRef(in, sink, base);
+            } else if (isLiteralWithPostBaseNameRef(b)) {
+                decodeLiteralWithPostBaseNameRef(in, sink, base);
+            } else if (isLiteral(b)) {
                 decodeLiteral(in, sink);
             } else {
                 throw UNKNOWN_TYPE;
             }
         }
+        if (requiredInsertCount > 0) {
+            final ByteBuf sectionAck = qpackDecoderStream.alloc().buffer();
```

Comment:
nit: we can do better in terms of initial size of the buffer here... 8 bytes should be enough for all values